2000 Arkansas Congressional Districts

KT0416838 opened the pull request:
## Redistricting requirements
In Arkansas, according to [NCSL Redistricting Law 2000](https://web.archive.org/web/20041216185957/https://www.senate.mn/departments/scr/redist/red2000/Tab5appx.htm), districts should:

1. be contiguous
1. have equal populations
1. maintain counties, cities, and established geographical boundaries, if possible
1. not dilute the voting strength of minority citizens

### Interpretation of requirements
We enforce a maximum population deviation of 0.5%, which is in line with the low deviation seen in past congressional district maps.
We apply a stronger county constraint, which is in line with the small number or absence of county splits observed in past congressional district maps.

## Data Sources
Data for ``Arkansas`` comes from the [ALARM Project's update](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/ZV5KF3) to [The Record of American Democracy](https://road.hmdc.harvard.edu/).

## Pre-processing Notes
No manual pre-processing decisions were necessary.

## Simulation Notes
We sample 20,000 districting plans for ``Arkansas`` across 5 independent runs of the SMC algorithm.
We then thinned the number of samples to 5,000. 
We use a stronger algorithmic county constraint to limit the number of county splits.


## Validation
<img width="3000" height="4500" alt="validation_20250814_1211" src="https://github.com/user-attachments/assets/cfa8ac49-dfdc-4fb5-a59f-0072368154cb" />


```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.42 to 0.82

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
        1.025         1.008         1.015         1.019         1.007 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.019         1.013         1.009         1.017         1.015 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
        1.006         1.018         1.015         1.008         1.008 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.020         1.016         1.016         1.008         1.010 
      vap_two county_splits   muni_splits           ndv           nrv 
        1.012         1.022         1.009         1.019         1.005 
      ndshare 
        1.009 

Sampling diagnostics for SMC run 1 of 5 (4,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     3,919 (98.0%)     12.8%        0.29 2,522 (100%)      9 
Split 2     3,424 (85.6%)     18.0%        0.35 2,510 ( 99%)      5 
Split 3     2,206 (55.1%)      7.8%        0.56 2,164 ( 86%)      4 
Resample      446 (11.2%)       NA%        1.31 1,761 ( 70%)     NA 

Sampling diagnostics for SMC run 2 of 5 (4,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     3,918 (98.0%)     11.4%        0.29 2,572 (102%)     10 
Split 2     3,430 (85.7%)     15.4%        0.35 2,514 ( 99%)      6 
Split 3     1,974 (49.3%)      7.6%        0.58 2,150 ( 85%)      4 
Resample       243 (6.1%)       NA%        1.31 1,612 ( 64%)     NA 

Sampling diagnostics for SMC run 3 of 5 (4,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     3,918 (98.0%)     11.6%        0.29 2,518 (100%)     10 
Split 2     3,431 (85.8%)     15.6%        0.35 2,497 ( 99%)      6 
Split 3     2,062 (51.5%)      7.6%        0.57 2,147 ( 85%)      4 
Resample       258 (6.4%)       NA%        1.32 1,637 ( 65%)     NA 

Sampling diagnostics for SMC run 4 of 5 (4,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     3,917 (97.9%)      7.5%        0.29 2,526 (100%)     15 
Split 2     3,412 (85.3%)     10.7%        0.36 2,487 ( 98%)      9 
Split 3     2,223 (55.6%)      6.3%        0.57 2,183 ( 86%)      5 
Resample      472 (11.8%)       NA%        1.31 1,762 ( 70%)     NA 

Sampling diagnostics for SMC run 5 of 5 (4,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     3,920 (98.0%)     12.6%        0.28 2,541 (100%)      9 
Split 2     3,416 (85.4%)     15.8%        0.37 2,505 ( 99%)      6 
Split 3     1,973 (49.3%)      7.7%        0.59 2,152 ( 85%)      4 
Resample       234 (5.9%)       NA%        1.33 1,582 ( 63%)     NA 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of unique
plans. R-hat values for summary statistics should be between 1 and 1.05.
```

## Checklist

- [x] I have followed the [instructions](https://github.com/alarm-redist/fifty-states/blob/main/CONTRIBUTING.md)
- [x] I have updated the [tracker](https://docs.google.com/spreadsheets/d/1k_tYLoE49W_DCK1tcWbouoYZFI9WD76oayEt5TOmJg4/edit#gid=453387933)
- [x] All `TODO` lines from the template code have been removed
- [x] I have merged in the main branch and then recalculated summary statistics
- [x] I have run `enforce_style()` to format my code
- [x] The documentation copied above is up-to-date 
- [x] There are no data files in this pull request
- [x] None of the file output paths (for the `redist_map` and `redist_plans` objects, and summary statistics) have been edited

@christopherkenny

christopherkenny commented:
@KT0416838, I'm going to hold off on this for now. Have you looked into the background here? The bill itself is explicit that it's a county-level plan, but I don't know if that's a convenience or general practice.

KT0416838 commented:
Thank you for your feedback. I tested higher county split constraints in 0.5 increments. Strength = 2 gave the best balance: the proportion of plans with 3 county splits decreased, and the proportion with 1 split increased, bringing results closer to the enacted plan. No warnings appeared in the summary. At strength = 2.5, county splits matched the enacted plan even more closely, but many R-hats exceeded 1.05; strength = 3 had even more problematic R-hats. If you also think that strength = 2 is optimal, I’ll update the pull request accordingly.

Strength = 2
<img width="3000" height="4500" alt="validation_20250808_2045" src="https://github.com/user-attachments/assets/29c24ce3-5cd3-43af-9b16-2dbaef874234" />
```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.40 to 0.77

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
        1.010         1.009         1.008         1.029         1.014 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.006         1.011         1.007         1.008         1.008 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
        1.010         1.009         1.007         1.008         1.007 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.008         1.008         1.008         1.011         1.012 
      vap_two county_splits   muni_splits           ndv           nrv 
        1.012         1.013         1.007         1.007         1.012 
      ndshare 
        1.008 

Sampling diagnostics for SMC run 1 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,961 (98.0%)     12.1%        0.28 1,276 (101%)      9 
Split 2     1,725 (86.3%)     13.9%        0.35 1,223 ( 97%)      7 
Split 3       910 (45.5%)      7.7%        0.58 1,104 ( 87%)      4 
Resample       111 (5.6%)       NA%        1.39   734 ( 58%)     NA 

Sampling diagnostics for SMC run 2 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,959 (97.9%)     11.3%        0.29 1,271 (101%)     10 
Split 2     1,728 (86.4%)     15.5%        0.35 1,249 ( 99%)      6 
Split 3     1,120 (56.0%)      7.8%        0.55 1,072 ( 85%)      4 
Resample       195 (9.8%)       NA%        1.25   893 ( 71%)     NA 

Sampling diagnostics for SMC run 3 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,961 (98.0%)     11.2%        0.28 1,260 (100%)     10 
Split 2     1,724 (86.2%)     15.5%        0.35 1,229 ( 97%)      6 
Split 3       982 (49.1%)      7.7%        0.58 1,094 ( 87%)      4 
Resample       108 (5.4%)       NA%        1.35   798 ( 63%)     NA 

Sampling diagnostics for SMC run 4 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,960 (98.0%)      7.7%        0.29 1,258 (100%)     15 
Split 2     1,715 (85.8%)     10.9%        0.34 1,235 ( 98%)      9 
Split 3     1,101 (55.1%)      5.4%        0.55 1,093 ( 86%)      6 
Resample      220 (11.0%)       NA%        1.31   870 ( 69%)     NA 

Sampling diagnostics for SMC run 5 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,958 (97.9%)     12.6%        0.29 1,266 (100%)      9 
Split 2     1,728 (86.4%)     15.7%        0.34 1,238 ( 98%)      6 
Split 3     1,124 (56.2%)      7.7%        0.55 1,109 ( 88%)      4 
Resample      217 (10.9%)       NA%        1.27   885 ( 70%)     NA 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of
unique plans. R-hat values for summary statistics should be between 1 and 1.05.
```

Strength = 2.5
<img width="3000" height="4500" alt="validation_20250808_2058" src="https://github.com/user-attachments/assets/9facec68-9dfd-4ecd-8549-b2ad9a521b57" />
```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.45 to 0.80

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
        1.040         1.025         1.029         1.015       ❌1.094 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.011       ❌1.071       ❌1.085       ❌1.082       ❌1.111 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
        1.039         1.014       ❌1.106       ❌1.065       ❌1.065 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.017       ❌1.055       ❌1.106         1.045         1.039 
      vap_two county_splits   muni_splits           ndv           nrv 
      ❌1.076       ❌1.072       ❌1.053       ❌1.094         1.028 
      ndshare 
        1.040 
✖ WARNING: SMC runs have not converged.

Sampling diagnostics for SMC run 1 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,960 (98.0%)     12.2%        0.28 1,279 (101%)      9    
Split 2     1,662 (83.1%)     18.4%        0.35 1,264 (100%)      5    
Split 3       710 (35.5%)      8.8%        0.60 1,077 ( 85%)      3    
Resample        64 (3.2%)       NA%        1.58   618 ( 49%)     NA  * 

Sampling diagnostics for SMC run 2 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,958 (97.9%)     11.5%        0.29 1,278 (101%)     10    
Split 2     1,650 (82.5%)     15.6%        0.36 1,249 ( 99%)      6    
Split 3       645 (32.2%)      7.7%        0.59 1,044 ( 83%)      4    
Resample        55 (2.8%)       NA%        1.61   560 ( 44%)     NA  * 

Sampling diagnostics for SMC run 3 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,959 (98.0%)     11.2%        0.29 1,279 (101%)     10    
Split 2     1,687 (84.3%)     15.4%        0.35 1,244 ( 98%)      6    
Split 3       581 (29.0%)      7.2%        0.57 1,036 ( 82%)      4    
Resample        36 (1.8%)       NA%        1.63   520 ( 41%)     NA  * 

Sampling diagnostics for SMC run 4 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,960 (98.0%)      7.6%        0.29 1,262 (100%)     15    
Split 2     1,641 (82.0%)     10.8%        0.35 1,267 (100%)      9    
Split 3       767 (38.3%)      3.0%        0.67 1,047 ( 83%)     11    
Resample        58 (2.9%)       NA%        1.51   655 ( 52%)     NA  * 

Sampling diagnostics for SMC run 5 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,958 (97.9%)     12.8%        0.29 1,268 (100%)      9 
Split 2     1,654 (82.7%)     15.8%        0.34 1,241 ( 98%)      6 
Split 3       876 (43.8%)      7.7%        0.63 1,070 ( 85%)      4 
Resample       196 (9.8%)       NA%        1.59   720 ( 57%)     NA 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of
unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low
plan diversity or bottlenecks as well, address those issues first.
• (*) Bottlenecks found: Consider weakening or removing constraints, or
increasing the population tolerance. If the acceptance rate drops quickly in the
final splits, try increasing `pop_temper` by 0.01. If the weight variance (Log
wgt. sd) increases steadily or is particularly large for the "Resample" step,
consider increasing `seq_alpha`. To visualize what geographic areas may be
causing problems, try running the following code. Highlighted areas are those
that may be causing the bottleneck.
```

Strength =3
<img width="3000" height="4500" alt="validation_20250808_2052" src="https://github.com/user-attachments/assets/eeea2ebe-01ed-4268-98c0-47f7188b3885" />
```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.38 to 0.77

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
        1.029       ❌1.063         1.022         1.027       ❌1.109 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.022       ❌1.055       ❌1.061       ❌1.061       ❌1.062 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
       ❌1.05         1.032         1.034       ❌1.061       ❌1.063 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.026       ❌1.062       ❌1.061         1.045         1.028 
      vap_two county_splits   muni_splits           ndv           nrv 
      ❌1.059         1.007         1.040         1.033       ❌1.063 
      ndshare 
      ❌1.068 
✖ WARNING: SMC runs have not converged.

Sampling diagnostics for SMC run 1 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,959 (97.9%)     12.5%        0.29 1,261 (100%)      9    
Split 2     1,614 (80.7%)     18.7%        0.34 1,233 ( 98%)      5    
Split 3       489 (24.5%)      7.8%        0.65 1,046 ( 83%)      4    
Resample        41 (2.1%)       NA%        1.95   429 ( 34%)     NA  * 

Sampling diagnostics for SMC run 2 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,960 (98.0%)     11.4%        0.28 1,259 (100%)     10    
Split 2     1,596 (79.8%)     15.6%        0.36 1,231 ( 97%)      6    
Split 3       492 (24.6%)      7.7%        0.64 1,065 ( 84%)      4    
Resample        46 (2.3%)       NA%        1.93   426 ( 34%)     NA  * 

Sampling diagnostics for SMC run 3 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,960 (98.0%)     11.1%        0.28 1,266 (100%)     10    
Split 2     1,613 (80.7%)     15.4%        0.35 1,252 ( 99%)      6    
Split 3       484 (24.2%)      7.4%        0.56 1,064 ( 84%)      4    
Resample        29 (1.4%)       NA%        1.85   439 ( 35%)     NA  * 

Sampling diagnostics for SMC run 4 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,960 (98.0%)      7.7%        0.28 1,273 (101%)     15    
Split 2     1,594 (79.7%)     11.9%        0.35 1,256 ( 99%)      8    
Split 3       505 (25.2%)      6.4%        0.66 1,049 ( 83%)      5    
Resample        47 (2.3%)       NA%        1.92   436 ( 34%)     NA  * 

Sampling diagnostics for SMC run 5 of 5 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k    
Split 1     1,960 (98.0%)     12.5%        0.28 1,269 (100%)      9    
Split 2     1,617 (80.9%)     15.3%        0.35 1,244 ( 98%)      6    
Split 3       503 (25.2%)      8.1%        0.62 1,051 ( 83%)      4    
Resample        21 (1.1%)       NA%        1.87   459 ( 36%)     NA  * 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of
unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low
plan diversity or bottlenecks as well, address those issues first.
• (*) Bottlenecks found: Consider weakening or removing constraints, or
increasing the population tolerance. If the acceptance rate drops quickly in the
final splits, try increasing `pop_temper` by 0.01. If the weight variance (Log
wgt. sd) increases steadily or is particularly large for the "Resample" step,
consider increasing `seq_alpha`. To visualize what geographic areas may be
causing problems, try running the following code. Highlighted areas are those
that may be causing the bottleneck.
```

kosukeimai commented:
Hmmm...  Still not great convergence.  @taransamarth Would you mind trying this with the new algorithm?

christopherkenny commented:
You may also want to update the number of samples per run here. That should help the higher weights function.

KT0416838 commented:
I increased `nsims` to 1.6e4.

- Strength = 2.5: Problematic R-hats dropped from 14 to 1.
- Strength = 3: Problematic R-hats dropped from 14 to 11.

I stopped at 1.6e4 because when I tried 2e4 for strength = 2.5, the number of problematic R-hats increased to 2. My impression is that further improvement would require a substantial increase in `nsims`, and I wasn’t sure if that would be appropriate at this stage.
If you think it’s worth pursuing, I can increase `nsims` further. I’m also interested in trying the new algorithm if it’s feasible for me to run.


1. Strength = 2.5, nsims = 1.6e4, runs = 5 — best results so far.

<img width="505" height="256" alt="Screenshot 2025-08-11 at 7 48 04 PM" src="https://github.com/user-attachments/assets/fd552faf-fd39-465b-956e-b9f1976c102b" />

```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.38 to 0.77

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
        1.009         1.020         1.020         1.019         1.038 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.027         1.011         1.015         1.010         1.009 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
        1.021         1.038         1.028         1.027         1.014 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.033         1.010         1.009         1.021         1.015 
      vap_two county_splits   muni_splits           ndv           nrv 
        1.011       ❌1.054         1.020         1.008         1.017 
      ndshare 
        1.011 
✖ WARNING: SMC runs have not converged.

Sampling diagnostics for SMC run 1 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,679 (98.0%)     12.5%        0.28 10,082 (100%)      9    
Split 2    13,247 (82.8%)      7.5%        0.36  9,962 ( 98%)     13    
Split 3     5,444 (34.0%)      4.2%        0.62  8,541 ( 84%)      8    
Resample       389 (2.4%)       NA%        1.60  4,678 ( 46%)     NA  * 

Sampling diagnostics for SMC run 2 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,674 (98.0%)     11.5%        0.29 10,119 (100%)     10    
Split 2    13,266 (82.9%)     13.9%        0.35  9,954 ( 98%)      7    
Split 3     5,856 (36.6%)      7.8%        0.60  8,598 ( 85%)      4    
Resample       537 (3.4%)       NA%        1.59  4,949 ( 49%)     NA  * 

Sampling diagnostics for SMC run 3 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,679 (98.0%)     11.4%        0.28 10,075 (100%)     10    
Split 2    13,261 (82.9%)     15.8%        0.35  9,980 ( 99%)      6    
Split 3     5,793 (36.2%)      4.2%        0.59  8,608 ( 85%)      8    
Resample       520 (3.3%)       NA%        1.58  4,951 ( 49%)     NA  * 

Sampling diagnostics for SMC run 4 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,675 (98.0%)      7.6%        0.29 10,180 (101%)     15    
Split 2    13,194 (82.5%)     10.9%        0.36  9,963 ( 99%)      9    
Split 3     6,139 (38.4%)      6.5%        0.61  8,522 ( 84%)      5    
Resample       558 (3.5%)       NA%        1.59  5,204 ( 51%)     NA  * 

Sampling diagnostics for SMC run 5 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,677 (98.0%)     12.5%        0.29 10,093 (100%)      9    
Split 2    13,222 (82.6%)     18.2%        0.36 10,000 ( 99%)      5    
Split 3     5,782 (36.1%)      7.8%        0.62  8,540 ( 84%)      4    
Resample       523 (3.3%)       NA%        1.61  4,934 ( 49%)     NA  * 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of
unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low
plan diversity or bottlenecks as well, address those issues first.
• (*) Bottlenecks found: Consider weakening or removing constraints, or
increasing the population tolerance. If the acceptance rate drops quickly in the
final splits, try increasing `pop_temper` by 0.01. If the weight variance (Log
wgt. sd) increases steadily or is particularly large for the "Resample" step,
consider increasing `seq_alpha`. To visualize what geographic areas may be
causing problems, try running the following code. Highlighted areas are those
that may be causing the bottleneck.
```

2. Strength = 3, nsims = 1.6e4, runs = 5 — partial improvement, still 11 problematic R-hats.

<img width="571" height="283" alt="Screenshot 2025-08-11 at 7 49 21 PM" src="https://github.com/user-attachments/assets/57132fc6-28c3-489e-8525-b41ad2d79210" />


```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.42 to 0.79

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
      ❌1.053       ❌1.062       ❌1.093         1.048       ❌1.083 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.046       ❌1.081         1.035         1.040         1.041 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
        1.048         1.042         1.046         1.047         1.034 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.032         1.045         1.041       ❌1.073       ❌1.126 
      vap_two county_splits   muni_splits           ndv           nrv 
      ❌1.059         1.028       ❌1.081       ❌1.065         1.041 
      ndshare 
      ❌1.069 
✖ WARNING: SMC runs have not converged.

Sampling diagnostics for SMC run 1 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,677 (98.0%)     12.4%        0.28 10,056 ( 99%)      9    
Split 2    12,865 (80.4%)     12.3%        0.35  9,975 ( 99%)      8    
Split 3     3,638 (22.7%)      6.4%        0.63  8,426 ( 83%)      5    
Resample       204 (1.3%)       NA%        1.87  3,375 ( 33%)     NA  * 

Sampling diagnostics for SMC run 2 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,675 (98.0%)     11.2%        0.29 10,132 (100%)     10    
Split 2    12,839 (80.2%)     15.6%        0.35  9,986 ( 99%)      6    
Split 3     3,205 (20.0%)      7.7%        0.62  8,414 ( 83%)      4    
Resample       152 (0.9%)       NA%        1.90  2,947 ( 29%)     NA  * 

Sampling diagnostics for SMC run 3 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,676 (98.0%)     11.2%        0.29 10,116 (100%)     10    
Split 2    12,840 (80.2%)     15.7%        0.35 10,010 ( 99%)      6    
Split 3     3,379 (21.1%)      4.8%        0.61  8,494 ( 84%)      7    
Resample       116 (0.7%)       NA%        1.87  3,154 ( 31%)     NA  * 

Sampling diagnostics for SMC run 4 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,676 (98.0%)      7.7%        0.29 10,130 (100%)     15    
Split 2    12,806 (80.0%)     10.9%        0.35  9,973 ( 99%)      9    
Split 3     3,708 (23.2%)      5.5%        0.63  8,431 ( 83%)      6    
Resample       181 (1.1%)       NA%        1.88  3,392 ( 34%)     NA  * 

Sampling diagnostics for SMC run 5 of 5 (16,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    15,677 (98.0%)     12.6%        0.29 10,089 (100%)      9    
Split 2    12,765 (79.8%)     18.1%        0.35  9,991 ( 99%)      5    
Split 3     3,529 (22.1%)      4.8%        0.62  8,506 ( 84%)      7    
Resample       148 (0.9%)       NA%        1.83  3,341 ( 33%)     NA  * 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of
unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low
plan diversity or bottlenecks as well, address those issues first.
• (*) Bottlenecks found: Consider weakening or removing constraints, or
increasing the population tolerance. If the acceptance rate drops quickly in the
final splits, try increasing `pop_temper` by 0.01. If the weight variance (Log
wgt. sd) increases steadily or is particularly large for the "Resample" step,
consider increasing `seq_alpha`. To visualize what geographic areas may be
causing problems, try running the following code. Highlighted areas are those
that may be causing the bottleneck.
```

3. Strength = 2.5, nsims = 2e4, runs = 5 — slightly worse than 1.6e4, so I stopped here.

<img width="509" height="253" alt="Screenshot 2025-08-11 at 7 50 09 PM" src="https://github.com/user-attachments/assets/9154261c-bfcf-43b8-992c-581d6765f583" />

```
SMC: 5,000 sampled plans of 4 districts on 2,805 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.5
`pop_temper`=0

Plan diversity 80% range: 0.40 to 0.77

R-hat values for summary statistics:
  pop_overlap     total_vap      plan_dev     comp_edge   comp_polsby 
        1.015         1.045         1.033         1.014         1.022 
    pop_other       pop_two      pop_aian     pop_white     pop_black 
        1.036         1.045         1.046         1.045         1.047 
     pop_hisp     pop_asian      pop_nhpi      vap_hisp      vap_aian 
        1.036         1.014         1.010         1.042         1.047 
    vap_asian     vap_white     vap_black      vap_nhpi     vap_other 
        1.018         1.046         1.047         1.025       ❌1.059 
      vap_two county_splits   muni_splits           ndv           nrv 
        1.047         1.015         1.010         1.048         1.043 
      ndshare 
      ❌1.055 
✖ WARNING: SMC runs have not converged.

Sampling diagnostics for SMC run 1 of 5 (20,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    19,598 (98.0%)     12.5%        0.28 12,630 (100%)      9    
Split 2    16,593 (83.0%)     15.7%        0.35 12,491 ( 99%)      6    
Split 3     6,648 (33.2%)      7.7%        0.60 10,726 ( 85%)      4    
Resample       352 (1.8%)       NA%        1.59  5,797 ( 46%)     NA  * 

Sampling diagnostics for SMC run 2 of 5 (20,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    19,595 (98.0%)     11.2%        0.29 12,596 (100%)     10    
Split 2    16,550 (82.7%)     15.8%        0.35 12,443 ( 98%)      6    
Split 3     6,908 (34.5%)      7.7%        0.60 10,655 ( 84%)      4    
Resample       400 (2.0%)       NA%        1.58  6,008 ( 48%)     NA  * 

Sampling diagnostics for SMC run 3 of 5 (20,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    19,594 (98.0%)     11.4%        0.29 12,689 (100%)     10    
Split 2    16,517 (82.6%)     15.7%        0.36 12,462 ( 99%)      6    
Split 3     7,037 (35.2%)      7.6%        0.60 10,716 ( 85%)      4    
Resample       447 (2.2%)       NA%        1.56  6,146 ( 49%)     NA  * 

Sampling diagnostics for SMC run 4 of 5 (20,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    19,598 (98.0%)      7.6%        0.28 12,595 (100%)     15    
Split 2    16,576 (82.9%)     10.9%        0.35 12,537 ( 99%)      9    
Split 3     6,529 (32.6%)      6.4%        0.60 10,704 ( 85%)      5    
Resample       402 (2.0%)       NA%        1.60  5,688 ( 45%)     NA  * 

Sampling diagnostics for SMC run 5 of 5 (20,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd   Max. unique Est. k    
Split 1    19,601 (98.0%)     12.5%        0.28 12,701 (100%)      9    
Split 2    16,606 (83.0%)     18.2%        0.35 12,484 ( 99%)      5    
Split 3     6,577 (32.9%)      7.7%        0.60 10,723 ( 85%)      4    
Resample       460 (2.3%)       NA%        1.62  5,661 ( 45%)     NA  * 

•  Watch out for low effective samples, very low acceptance rates (less than 1%),
large std. devs. of the log weights (more than 3 or so), and low numbers of
unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• SMC convergence: Increase the number of samples. If you are experiencing low
plan diversity or bottlenecks as well, address those issues first.
• (*) Bottlenecks found: Consider weakening or removing constraints, or
increasing the population tolerance. If the acceptance rate drops quickly in the
final splits, try increasing `pop_temper` by 0.01. If the weight variance (Log
wgt. sd) increases steadily or is particularly large for the "Resample" step,
consider increasing `seq_alpha`. To visualize what geographic areas may be
causing problems, try running the following code. Highlighted areas are those
that may be causing the bottleneck.
```

kosukeimai commented:
It seems to me that our new algorithm might help.  @taransamarth Can you help out here?
